src/program.hpp

```diff
@@ -634,8 +634,19 @@ struct cvk_program : public _cl_program, api_object<object_magic::program> {
 private:
     void do_build();
     std::string prepare_build_options(const cvk_device* device) const;
-    CHECK_RETURN cl_build_status compile_source(const cvk_device* device);
-    CHECK_RETURN cl_build_status link();
+    CHECK_RETURN cl_build_status build_source(const cvk_device* device);
+
+#if COMPILER_AVAILABLE
+#ifndef CLSPV_ONLINE_COMPILER
+    CHECK_RETURN cl_build_status
+    build_source_offline(bool build_to_ir, bool build_from_il,
+                         std::string& build_options, std::string& tmp_folder);
+#else
+    CHECK_RETURN cl_build_status build_source_online(
+        bool build_to_ir, bool build_from_il, std::string& build_options);
+#endif
+#endif
```

Comment:
I like how you split what was becoming a monster into online and offline functions but I don't think `build_source*` is a good name any more (the naming problem existed before this change).